File: app/lib/data/metadata.dart
import 'dart:async';
import 'dart:convert';
import 'dart:typed_data';

import 'package:http/http.dart' as http;

import '../add_book/isbn.dart';

/// One edition/work found by an online metadata search. The same shape is
/// produced by every source (Open Library, Google Books), so the rest of the
/// app doesn't care where a result came from.
class BookSearchResult {
  const BookSearchResult({
    required this.workKey,
    required this.title,
    this.subtitle,
    this.authors = const [],
    this.firstPublishYear,
    this.isbn,
    this.coverId,
    this.coverUrl,
    this.description,
    this.subjects = const [],
    this.publisher,
    this.pageCount,
  });

  /// Open Library work key, e.g. "/works/OL45883W". Empty for other sources;
  /// used to fetch the full description lazily (see [OpenLibraryClient]).
  final String workKey;
  final String title;
  final String? subtitle;
  final List<String> authors;
  final int? firstPublishYear;
  final String? isbn;

  /// Open Library cover id (its covers are addressed by numeric id).
  final int? coverId;

  /// A ready-made cover image URL, used by sources that hand back a link
  /// rather than an id (Google Books). Takes precedence over [coverId].
  final Uri? coverUrl;

  /// Description, when the source returns it inline (Google Books does;
  /// Open Library needs a second request — see [MetadataService.descriptionOf]).
  final String? description;

  final List<String> subjects;
  final String? publisher;
  final int? pageCount;

  String get authorLine => authors.isEmpty ? 'Unknown author' : authors.join(', ');

  /// Small cover for search result lists.
  Uri? get thumbnailUrl {
    if (coverId != null) {
      return Uri.parse('https://covers.openlibrary.org/b/id/$coverId-M.jpg');
    }
    return coverUrl;
  }

  /// Full-size cover to store with the book, or null if none is known.
  Uri? get largeCoverUrl {
    if (coverId != null) {
      return Uri.parse('https://covers.openlibrary.org/b/id/$coverId-L.jpg');
    }
    return coverUrl;
  }

  /// Open Library's *books* API (`/api/books?jscmd=data`), which answers from
  /// the edition records rather than the search index — see
  /// [OpenLibraryClient.lookupEdition] for why that is a different question.
  /// The same result under the barcode that was scanned. A lookup may have
  /// succeeded against the ISBN-10 form, but the book is stored under the
  /// 13-digit one — that is the key duplicate detection and search use.
  BookSearchResult withIsbn(String isbn13) => BookSearchResult(
        workKey: workKey,
        title: title,
        subtitle: subtitle,
        authors: authors,
        firstPublishYear: firstPublishYear,
        isbn: isbn13,
        coverId: coverId,
        coverUrl: coverUrl,
        description: description,
        subjects: subjects,
        publisher: publisher,
        pageCount: pageCount,
      );

  factory BookSearchResult.fromOpenLibraryData(
    Map<String, dynamic> data, {
    required String isbn,
  }) {
    List<String> names(dynamic v) => v is List
        ? [
            for (final e in v)
              if (e is Map && e['name'] is String) e['name'] as String,
          ]
        : const [];
    final cover = data['cover'];
    final coverUrl = cover is Map ? (cover['large'] ?? cover['medium']) : null;
    // "1990" or "March 1990" or "1990-03-01": the year is the part every one
    // of those has, and the only part this app shows.
    final year = RegExp(r'\d{4}').firstMatch('${data['publish_date'] ?? ''}');
    return BookSearchResult(
      workKey: '',
      title: (data['title'] as String?) ?? '',
      subtitle: data['subtitle'] as String?,
      authors: names(data['authors']),
      firstPublishYear: year == null ? null : int.tryParse(year.group(0)!),
      isbn: isbn,
      coverUrl: coverUrl is String ? Uri.tryParse(coverUrl) : null,
      subjects: names(data['subjects']),
      publisher: names(data['publishers']).firstOrNull,
      pageCount: (data['number_of_pages'] as num?)?.toInt(),
    );
  }

  factory BookSearchResult.fromOpenLibraryDoc(Map<String, dynamic> doc) {
    List<String> strings(dynamic v) =>
        v is List ? v.whereType<String>().toList() : const [];
    return BookSearchResult(
      workKey: doc['key'] as String? ?? '',
      title: doc['title'] as String? ?? '',
      subtitle: doc['subtitle'] as String?,
      authors: strings(doc['author_name']),
      firstPublishYear: doc['first_publish_year'] as int?,
      isbn: strings(doc['isbn']).firstOrNull,
      coverId: doc['cover_i'] as int?,
      subjects: strings(doc['subject']),
      publisher: strings(doc['publisher']).firstOrNull,
      pageCount: doc['number_of_pages_median'] as int?,
    );
  }

  factory BookSearchResult.fromGoogleVolume(Map<String, dynamic> volume) {
    final info =
        volume['volumeInfo'] as Map<String, dynamic>? ?? const {};
    List<String> strings(dynamic v) =>
        v is List ? v.whereType<String>().toList() : const [];

    // Published date is "2003", "2003-05" or "2003-05-17"; take the year.
    int? year;
    final published = info['publishedDate'] as String?;
    if (published != null && published.length >= 4) {
      year = int.tryParse(published.substring(0, 4));
    }

    // Prefer an ISBN-13, else whatever identifier is present.
    String? isbn;
    final ids = info['industryIdentifiers'] as List? ?? const [];
    for (final id in ids.whereType<Map<String, dynamic>>()) {
      final value = id['identifier'] as String?;
      if (value == null) continue;
      isbn ??= value;
      if (id['type'] == 'ISBN_13') {
        isbn = value;
        break;
      }
    }

    final links = info['imageLinks'] as Map<String, dynamic>? ?? const {};
    final thumb = (links['thumbnail'] ?? links['smallThumbnail']) as String?;

    return BookSearchResult(
      workKey: '',
      title: info['title'] as String? ?? '',
      subtitle: info['subtitle'] as String?,
      authors: strings(info['authors']),
      firstPublishYear: year,
      isbn: isbn,
      // Google serves covers over http by default; force https for the CSP.
      coverUrl: thumb == null
          ? null
          : Uri.parse(thumb.replaceFirst('http://', 'https://')),
      description: info['description'] as String?,
      subjects: strings(info['categories']),
      publisher: info['publisher'] as String?,
      pageCount: info['pageCount'] as int?,
    );
  }
}

/// Client for the Open Library APIs (https://openlibrary.org/developers/api).
/// Free, no API key.
class OpenLibraryClient {
  OpenLibraryClient([http.Client? client]) : _http = client ?? http.Client();

  final http.Client _http;

  /// Searches works by free text (title, author, or ISBN all work).
  Future<List<BookSearchResult>> search(String query) async {
    final uri = Uri.https('openlibrary.org', '/search.json', {
      'q': query,
      'fields': 'key,title,subtitle,author_name,first_publish_year,isbn,'
          'cover_i,subject,publisher,number_of_pages_median',
      'limit': '20',
    });
    final res = await _http.get(uri);
    if (res.statusCode != 200) {
      throw Exception('Open Library search failed (HTTP ${res.statusCode})');
    }
    final docs = (jsonDecode(res.body)['docs'] as List? ?? const [])
        .whereType<Map<String, dynamic>>();
    return [
      for (final doc in docs)
        if ((doc['title'] as String?)?.isNotEmpty ?? false)
          BookSearchResult.fromOpenLibraryDoc(doc),
    ];
  }

  /// Looks a book up by its ISBN specifically (plan 5 #16).
  ///
  /// `isbn:<n>` rather than the bare number as free text: a plain numeric query
  /// also matches works that merely *mention* the digits, which for a barcode
  /// scan means confidently attaching the wrong book.
  Future<List<BookSearchResult>> searchByIsbn(String isbn13) =>
      search('isbn:$isbn13');

  /// One edition, straight from the edition record.
  ///
  /// A different question from [searchByIsbn], and that is the point: the
  /// search index only knows editions that made it into a *work*, while this
  /// answers from the edition itself. A book catalogued by a library or added
  /// by an importer — which is most of the ones a barcode finds nothing for —
  /// is often present here and absent there.
  Future<BookSearchResult?> lookupEdition(String isbn) async {
    final uri = Uri.https('openlibrary.org', '/api/books', {
      'bibkeys': 'ISBN:$isbn',
      'format': 'json',
      'jscmd': 'data',
    });
    final res = await _http.get(uri);
    if (res.statusCode != 200) return null;
    final body = jsonDecode(res.body);
    if (body is! Map) return null;
    final data = body['ISBN:$isbn'];
    if (data is! Map<String, dynamic>) return null;
    final result = BookSearchResult.fromOpenLibraryData(data, isbn: isbn);
    return result.title.isEmpty ? null : result;
  }

  /// The work's description is not in search results; fetch it separately.
  Future<String?> fetchDescription(String workKey) async {
    if (workKey.isEmpty) return null;
    final res = await _http.get(Uri.https('openlibrary.org', '$workKey.json'));
    if (res.statusCode != 200) return null;
    final desc = (jsonDecode(res.body) as Map<String, dynamic>)['description'];
    if (desc is String) return desc;
    if (desc is Map && desc['value'] is String) return desc['value'] as String;
    return null;
  }
}

/// Client for the Google Books API (https://developers.google.com/books).
/// Free for public volume search, no API key required for basic queries.
class GoogleBooksClient {
  GoogleBooksClient([http.Client? client]) : _http = client ?? http.Client();

  final http.Client _http;

  /// Google Books' own ISBN-qualified search — the fallback for a scan when
  /// Open Library doesn't have the edition.
  Future<List<BookSearchResult>> searchByIsbn(String isbn13) =>
      search('isbn:$isbn13');

  Future<List<BookSearchResult>> search(String query) async {
    final uri = Uri.https('www.googleapis.com', '/books/v1/volumes', {
      'q': query,
      'maxResults': '20',
      'printType': 'books',
    });
    final res = await _http.get(uri);
    if (res.statusCode != 200) {
      throw Exception('Google Books search failed (HTTP ${res.statusCode})');
    }
    final items = (jsonDecode(res.body)['items'] as List? ?? const [])
        .whereType<Map<String, dynamic>>();
    return [
      for (final item in items)
        if ((((item['volumeInfo'] as Map?)?['title']) as String?)?.isNotEmpty ??
            false)
          BookSearchResult.fromGoogleVolume(item),
    ];
  }
}

/// Combines the metadata sources per DESIGN.md: query Open Library first
/// (free, no key), fall back to Google Books when it has nothing (or is down).
/// Downloading covers and resolving descriptions is source-agnostic.
class MetadataService {
  MetadataService({
    OpenLibraryClient? openLibrary,
    GoogleBooksClient? googleBooks,
    http.Client? client,
  })  : _openLibrary = openLibrary ?? OpenLibraryClient(client),
        _googleBooks = googleBooks ?? GoogleBooksClient(client),
        _http = client ?? http.Client();

  final OpenLibraryClient _openLibrary;
  final GoogleBooksClient _googleBooks;
  final http.Client _http;

  /// Open Library first; Google Books only if Open Library returns nothing or
  /// errors out — so a working primary source is never blocked by the fallback.
  Future<List<BookSearchResult>> search(String query) async {
    List<BookSearchResult> openLibraryResults = const [];
    try {
      openLibraryResults = await _openLibrary.search(query);
    } catch (_) {
      // Fall through to Google Books below.
    }
    if (openLibraryResults.isNotEmpty) return openLibraryResults;
    return _googleBooks.search(query);
  }

  /// The one book behind a scanned barcode, or null if nothing knows it
  /// (plan 5 #16).
  ///
  /// Reduced to a single result: a barcode identifies one edition, so a list of
  /// twenty would only be a worse confirm step. Returns null rather than
  /// throwing when nothing matches — "not found" is an ordinary outcome the
  /// caller answers with the manual form.
  ///
  /// **Why it asks so many times.** A book that "isn't in the database" usually
  /// is, under a different key. Each of these finds books the others don't:
  ///
  ///  * Open Library's *search* index — good coverage, but only editions that
  ///    were folded into a work.
  ///  * Google Books — a different catalogue entirely, and better on recent and
  ///    non-English printings.
  ///  * Open Library's *edition* record — the same library, a different index;
  ///    imported and library-catalogued editions live here and nowhere else.
  ///  * All three again against the **ISBN-10** form. Records made before 2007
  ///    are keyed by it, and plenty were never re-indexed, so the ten-digit
  ///    form of the barcode on the back of the book finds what the barcode
  ///    itself does not.
  ///
  /// The order is cheapest-and-likeliest first, and it stops at the first hit,
  /// so the common case is still one request.
  Future<BookSearchResult?> lookupByIsbn(String isbn13) async {
    final forms = <String>[isbn13, ?isbn13To10(isbn13)];
    for (final isbn in forms) {
      for (final attempt in <Future<List<BookSearchResult>> Function()>[
        () => _openLibrary.searchByIsbn(isbn),
        () => _googleBooks.searchByIsbn(isbn),
        () async {
          final edition = await _openLibrary.lookupEdition(isbn);
          return edition == null ? const [] : [edition];
        },
      ]) {
        try {
          // A source that hangs must not hold the scanner: six catalogues with
          // no deadline is a scan that never comes back.
          final results = await attempt().timeout(const Duration(seconds: 8));
          // Whatever the source was asked about, record the *barcode* on the
          // book: it is the key the rest of the app dedupes and searches by.
          final hit = results.firstOrNull;
          if (hit != null) return hit.withIsbn(isbn13);
        } catch (_) {
          // A source that is down or rate-limiting is not an answer; ask the
          // next one rather than failing the scan.
        }
      }
    }
    return null;
  }

  /// The description for a chosen result: inline when the source supplied one
  /// (Google Books), otherwise fetched from Open Library by its work key.
  Future<String?> descriptionOf(BookSearchResult result) async {
    if (result.description != null) return result.description;
    return _openLibrary.fetchDescription(result.workKey);
  }

  /// Downloads the full-size cover for a result, or null if there is none /
  /// it fails.
  Future<Uint8List?> downloadCover(BookSearchResult result) async {
    final url = result.largeCoverUrl;
    if (url == null) return null;
    final res = await _http.get(url);
    if (res.statusCode != 200 || res.bodyBytes.isEmpty) return null;
    return res.bodyBytes;
  }
}

File: server/src/metadata.rs
//! Online book metadata search — the server-side twin of the app's
//! `metadata.dart`: query Open Library first, fall back to Google Books.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::error::{AppError, AppResult};

/// One edition/work found by a metadata search. Doubles as the body the console
/// posts back to add the chosen book.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BookSearchResult {
    #[serde(default)]
    pub work_key: String,
    pub title: String,
    #[serde(default)]
    pub subtitle: Option<String>,
    #[serde(default)]
    pub authors: Vec<String>,
    #[serde(default)]
    pub first_publish_year: Option<i64>,
    #[serde(default)]
    pub isbn: Option<String>,
    #[serde(default)]
    pub cover_id: Option<i64>,
    #[serde(default)]
    pub cover_url: Option<String>,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub subjects: Vec<String>,
    #[serde(default)]
    pub publisher: Option<String>,
    #[serde(default)]
    pub page_count: Option<i64>,
}

impl BookSearchResult {
    /// Full-size cover URL, preferring the Open Library cover id.
    pub fn large_cover_url(&self) -> Option<String> {
        match self.cover_id {
            Some(id) => Some(format!("https://covers.openlibrary.org/b/id/{id}-L.jpg")),
            None => self.cover_url.clone(),
        }
    }
}

/// Open Library first; Google Books only if it returns nothing or errors. A
/// source error is treated as "no results" so callers can offer to create a
/// custom book instead of failing.
pub async fn search(http: &reqwest::Client, query: &str) -> AppResult<Vec<BookSearchResult>> {
    let open_library = open_library_search(http, query).await.unwrap_or_default();
    if !open_library.is_empty() {
        return Ok(open_library);
    }
    Ok(google_books_search(http, query).await.unwrap_or_default())
}

/// The work's description (Open Library holds it separately from search).
pub async fn fetch_description(
    http: &reqwest::Client,
    work_key: &str,
) -> AppResult<Option<String>> {
    if work_key.is_empty() {
        return Ok(None);
    }
    let res = http
        .get(format!("https://openlibrary.org{work_key}.json"))
        .send()
        .await
        .map_err(net)?;
    if !res.status().is_success() {
        return Ok(None);
    }
    let body: Value = res.json().await.map_err(net)?;
    Ok(match body.get("description") {
        Some(Value::String(s)) => Some(s.clone()),
        Some(Value::Object(o)) => o.get("value").and_then(|v| v.as_str()).map(String::from),
        _ => None,
    })
}

/// Downloads the full-size cover for a result, or None if it has none / fails.
pub async fn download_cover(http: &reqwest::Client, result: &BookSearchResult) -> Option<Vec<u8>> {
    let url = result.large_cover_url()?;
    let res = http.get(url).send().await.ok()?;
    if !res.status().is_success() {
        return None;
    }
    let bytes = res.bytes().await.ok()?;
    (!bytes.is_empty()).then(|| bytes.to_vec())
}

async fn open_library_search(
    http: &reqwest::Client,
    query: &str,
) -> AppResult<Vec<BookSearchResult>> {
    let res = http
        .get("https://openlibrary.org/search.json")
        .query(&[
            ("q", query),
            (
                "fields",
                "key,title,subtitle,author_name,first_publish_year,isbn,cover_i,\
                 subject,publisher,number_of_pages_median",
            ),
            ("limit", "20"),
        ])
        .send()
        .await
        .map_err(net)?;
    if !res.status().is_success() {
        return Err(AppError::Internal(format!(
            "Open Library search failed (HTTP {})",
            res.status()
        )));
    }
    let body: Value = res.json().await.map_err(net)?;
    let docs = body.get("docs").and_then(Value::as_array);
    Ok(docs
        .map(|d| d.iter().filter_map(parse_open_library_doc).collect())
        .unwrap_or_default())
}

async fn google_books_search(
    http: &reqwest::Client,
    query: &str,
) -> AppResult<Vec<BookSearchResult>> {
    let res = http
        .get("https://www.googleapis.com/books/v1/volumes")
        .query(&[("q", query), ("maxResults", "20"), ("printType", "books")])
        .send()
        .await
        .map_err(net)?;
    if !res.status().is_success() {
        return Err(AppError::Internal(format!(
            "Google Books search failed (HTTP {})",
            res.status()
        )));
    }
    let body: Value = res.json().await.map_err(net)?;
    let items = body.get("items").and_then(Value::as_array);
    Ok(items
        .map(|i| i.iter().filter_map(parse_google_volume).collect())
        .unwrap_or_default())
}

fn parse_open_library_doc(doc: &Value) -> Option<BookSearchResult> {
    let title = str_field(doc, "title")?;
    if title.is_empty() {
        return None;
    }
    Some(BookSearchResult {
        work_key: str_field(doc, "key").unwrap_or_default(),
        title,
        subtitle: str_field(doc, "subtitle"),
        authors: str_array(doc, "author_name"),
        first_publish_year: doc.get("first_publish_year").and_then(Value::as_i64),
        isbn: str_array(doc, "isbn").into_iter().next(),
        cover_id: doc.get("cover_i").and_then(Value::as_i64),
        cover_url: None,
        description: None,
        subjects: str_array(doc, "subject"),
        publisher: str_array(doc, "publisher").into_iter().next(),
        page_count: doc.get("number_of_pages_median").and_then(Value::as_i64),
    })
}

fn parse_google_volume(item: &Value) -> Option<BookSearchResult> {
    let info = item.get("volumeInfo")?;
    let title = str_field(info, "title")?;
    if title.is_empty() {
        return None;
    }

    let year = str_field(info, "publishedDate")
        .and_then(|d| d.get(0..4).and_then(|y| y.parse::<i64>().ok()));

    // Prefer an ISBN-13, else whatever identifier is present.
    let mut isbn = None;
    if let Some(ids) = info.get("industryIdentifiers").and_then(Value::as_array) {
        for id in ids {
            if let Some(value) = str_field(id, "identifier") {
                isbn.get_or_insert(value.clone());
                if str_field(id, "type").as_deref() == Some("ISBN_13") {
                    isbn = Some(value);
                    break;
                }
            }
        }
    }

    let thumbnail = info
        .get("imageLinks")
        .and_then(|links| links.get("thumbnail").or_else(|| links.get("smallThumbnail")))
        .and_then(Value::as_str)
        .map(|s| s.replace("http://", "https://"));

    Some(BookSearchResult {
        work_key: String::new(),
        title,
        subtitle: str_field(info, "subtitle"),
        authors: str_array(info, "authors"),
        first_publish_year: year,
        isbn,
        cover_id: None,
        cover_url: thumbnail,
        description: str_field(info, "description"),
        subjects: str_array(info, "categories"),
        publisher: str_field(info, "publisher"),
        page_count: info.get("pageCount").and_then(Value::as_i64),
    })
}

fn net(e: reqwest::Error) -> AppError {
    AppError::Internal(format!("metadata source error: {e}"))
}

fn str_field(value: &Value, key: &str) -> Option<String> {
    value.get(key).and_then(Value::as_str).map(String::from)
}

fn str_array(value: &Value, key: &str) -> Vec<String> {
    value
        .get(key)
        .and_then(Value::as_array)
        .map(|a| a.iter().filter_map(Value::as_str).map(String::from).collect())
        .unwrap_or_default()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_an_open_library_doc() {
        let doc = serde_json::json!({
            "key": "/works/OL45883W",
            "title": "Dune",
            "author_name": ["Frank Herbert"],
            "first_publish_year": 1965,
            "isbn": ["9780441013593"],
            "cover_i": 12345,
            "number_of_pages_median": 412
        });
        let parsed = parse_open_library_doc(&doc).unwrap();
        assert_eq!(parsed.title, "Dune");
        assert_eq!(parsed.authors, vec!["Frank Herbert".to_string()]);
        assert_eq!(parsed.first_publish_year, Some(1965));
        assert_eq!(parsed.cover_id, Some(12345));
        assert_eq!(
            parsed.large_cover_url().as_deref(),
            Some("https://covers.openlibrary.org/b/id/12345-L.jpg")
        );
    }

    #[test]
    fn parses_a_google_volume() {
        let item = serde_json::json!({
            "volumeInfo": {
                "title": "Neuromancer",
                "authors": ["William Gibson"],
                "publishedDate": "1984-07-01",
                "industryIdentifiers": [
                    {"type": "ISBN_10", "identifier": "0441569595"},
                    {"type": "ISBN_13", "identifier": "9780441569595"}
                ],
                "pageCount": 271
            }
        });
        let parsed = parse_google_volume(&item).unwrap();
        assert_eq!(parsed.title, "Neuromancer");
        assert_eq!(parsed.first_publish_year, Some(1984));
        assert_eq!(parsed.isbn.as_deref(), Some("9780441569595"));
        assert_eq!(parsed.page_count, Some(271));
    }
}
